wgengine/netstack/link_endpoint.go

```diff
@@ -132,19 +156,117 @@ func (l *linkEndpoint) Drain() int {
 	return c
 }
 
-// NumQueued returns the number of packet queued for outbound.
+// NumQueued returns the number of packets queued for outbound.
 func (l *linkEndpoint) NumQueued() int {
 	return l.q.Num()
 }
 
-// InjectInbound injects an inbound packet. If the endpoint is not attached, the
-// packet is not delivered.
-func (l *linkEndpoint) InjectInbound(protocol tcpip.NetworkProtocolNumber, pkt *stack.PacketBuffer) {
+// rxChecksumOffload validates IPv4, TCP, and UDP header checksums in p,
+// returning an equivalent *stack.PacketBuffer if they are valid, otherwise nil.
+// The set of headers validated covers where gVisor would perform validation if
+// !stack.PacketBuffer.RXChecksumValidated, i.e. it satisfies
+// stack.CapabilityRXChecksumOffload.
+func rxChecksumOffload(p *packet.Parsed) *stack.PacketBuffer {
+	var (
+		pn        tcpip.NetworkProtocolNumber
+		csumStart int
+	)
+	buf := p.Buffer()
+
+	switch p.IPVersion {
+	case 4:
+		if len(buf) < 20 {
+			return nil
+		}
+		csumStart = int((buf[0] & 0x0F) * 4)
+		if csumStart < 20 || csumStart > 60 || len(buf) < csumStart {
+			return nil
+		}
+		if ^tun.Checksum(buf[:csumStart], 0) != 0 {
+			return nil
+		}
+		pn = header.IPv4ProtocolNumber
+	case 6:
+		if len(buf) < 40 {
+			return nil
+		}
+		csumStart = 40
+		pn = header.IPv6ProtocolNumber
+	}
+
+	if p.IPProto == ipproto.TCP || p.IPProto == ipproto.UDP {
+		lenForPseudo := len(buf) - csumStart
+		csum := tun.PseudoHeaderChecksum(
+			uint8(p.IPProto),
+			p.Src.Addr().AsSlice(),
+			p.Dst.Addr().AsSlice(),
+			uint16(lenForPseudo))
+		csum = tun.Checksum(buf[csumStart:], csum)
+		if ^csum != 0 {
+			return nil
+		}
+	}
+
+	packetBuf := stack.NewPacketBuffer(stack.PacketBufferOptions{
+		Payload: buffer.MakeWithData(bytes.Clone(buf)),
+	})
+	packetBuf.NetworkProtocolNumber = pn
+	// Setting this is not technically required. gVisor overrides where
+	// stack.CapabilityRXChecksumOffload is advertised from Capabilities().
+	// https://github.com/google/gvisor/blob/64c016c92987cc04dfd4c7b091ddd21bdad875f8/pkg/tcpip/stack/nic.go#L763
+	// This is also why we offload for all packets since we cannot signal this
+	// per-packet.
+	packetBuf.RXChecksumValidated = true
+	return packetBuf
+}
+
+func (l *linkEndpoint) injectInbound(p *packet.Parsed) {
 	l.mu.RLock()
 	d := l.dispatcher
 	l.mu.RUnlock()
 	if d != nil {
-		d.DeliverNetworkPacket(protocol, pkt)
+		pkt := rxChecksumOffload(p)
+		if pkt != nil {
+			d.DeliverNetworkPacket(pkt.NetworkProtocolNumber, pkt)
+			pkt.DecRef()
```

Comment:
This looks new; some spelunking around the gvisor codebase seems to indicate that it's correct; `pkg/tcpip/stack/gro/gro.go` calls `handlePacket` (which calls `gd.Dispatcher.DeliverNetworkPacket`), and then `pkt.DecRef`. Just confirming that this is intentional?

Reply:
yes, `rxChecksumOffload()` allocates `pkt` (ref count = 1), so we are responsible for decrementing here when we are done.

`d.DeliverNetworkPacket()` & `l.gro.Enqueue()` will increment ref count if they hold `pkt` past their return, e.g. GRO holds `pkt` in a bucket until later flush

Reply:
also, see the old `injectInbound()` in netstack.go. This is just shuffling the DecRef() over here.

Reply:
Ah yeah, I see that now; thanks!